src/cla/tcp.rs

```diff
@@ -0,0 +1,511 @@
+use log::{debug, error, info, warn};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::net::SocketAddr;
+use std::ops::DerefMut;
+use std::sync::Arc;
+use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
+use tokio::sync::mpsc::error::SendError;
+use tokio::task::JoinHandle;
+use tokio::time;
+//use std::net::TcpStream;
+use super::tcpcl::proto::*;
+use anyhow::bail;
+use bytes::Bytes;
+use thiserror::Error;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::TcpListener;
+use tokio::net::TcpStream;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::Duration;
+
+struct TcpSession {
+    sender_task: Sender<TcpClPacket>,
+    receiver_task: Receiver<TcpClPacket>,
+    bundle_queue: Receiver<(SocketAddr, ByteBuffer)>,
+    data_local: SessInitData,
+    data_remote: SessInitData,
+    last_tid: u64,
+}
+
+#[derive(Debug, Error)]
+enum SessionError {
+    #[error("error using internal channel")]
+    InternalChannel(#[from] SendError<TcpClPacket>),
+}
+
+impl TcpSession {
+    async fn run(mut self) {
+        loop {
+            tokio::select! {
+                received = self.receiver_task.recv() => {
+                    if let Some(packet) = received {
+                        debug!("Receiving bundle");
+                        match self.receive(packet).await {
+                            Ok(stop) => if stop { break; },
+                            Err(err) => error!("error while receiving: {:?}", err),
+                        }
+                    } else {
+                        break;
+                    }
+                },
+                bundle = self.bundle_queue.recv() =>  {
+                    if let Some(message) = bundle {
+                        debug!("Sending bundle");
+                        self.send(message).await;
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    async fn receive(&mut self, packet: TcpClPacket) -> Result<bool, SessionError> {
+        match packet {
+            TcpClPacket::SessTerm(data) => {
+                if !data.flags.contains(SessTermFlags::REPLY) {
+                    self.sender_task
+                        .send(TcpClPacket::SessTerm(SessTermData {
+                            flags: SessTermFlags::REPLY,
+                            reason: data.reason,
+                        }))
+                        .await?;
+                }
+                return Ok(true);
+            }
+            TcpClPacket::XferSeg(data) => {
+                if data.flags.contains(XferSegmentFlags::END)
+                    && !data.flags.contains(XferSegmentFlags::START)
+                {
+                    warn!("Received XferSeg with end flag, unexpected in idle mode");
+                }
+                if data.flags.is_empty() {
+                    warn!("Received XferSeg with no flags, unexpected in idle mode");
```

Comment:
shouldn't unexpected packets trigger a session termination? or at least stop processing the packet as a regular one?